synapse/storage/databases/main/registration.py

```diff
@@ -1082,6 +1084,21 @@ async def update_access_token_last_validated(self, token_id: int) -> None:
             desc="update_access_token_last_validated",
         )
 
+    async def mark_access_token_as_used(self, token_id: int) -> None:
+        """Mark the access token as used, which invalidates the old refresh token.
+
+        Args:
+            token_id: The ID of the access token to update.
+        Raises:
+            StoreError if there was a problem updating this.
+        """
+        await self.db_pool.simple_update_one(
+            "access_tokens",
+            {"id": token_id},
+            {"used": True},
+            desc="mark_access_token_as_used",
+        )
```

Comment:
note that `get_user_by_access_token` has a cache on it, so you'll have to invalidate the cache after updating the table.

I previously suggested joining to `user_ips` instead of modifying `access_tokens`: would that not be sufficient?